arthur/exts/directory/ldap.py

```diff
@@ -0,0 +1,416 @@
+- [PyDis ID Self-Service](<https://id.pydis.wtf/>)
+- [ModMail](https://modmail.pydis.wtf/)
+- Anti-Spam Message Deletions
+If you require data from the old system, please let us know and we can assist you in migrating it over.
+"""
+
+CREDENTIALS_SECTION = """
+## {title}
+
+To get started, you will need to login [here](<https://id.pydis.wtf/realms/pydis/account>   ) using the following credentials:
+
+- **Username:** `{username}@pydis.wtf`
+- **Password:** ||`{password}`||
+
+You will be prompted to reset your password after logging in.
+"""
+
+ELIGIBLE_MESSAGE = """
+Hi {mention}! You have roles that make you eligible for a new LDAP account. Please read the message above to get started!
+"""
+
+
+class LDAPSyncAction(StrEnum):
+    """Represents an action that will be performed against the LDAP directory."""
+
+    ADD = "add"
+    REMOVE = "remove"
+    KEEP = "keep"
+    CHANGE = "change"
+
+
+@dataclass
+class DiffedUser:
+    """Represents a user with their Discord and LDAP records, as well as if modification is required."""
+
+    discord_user: discord.Member
+    ldap_user: ldap.LDAPUser | None
+    groups: list[str]
+    action: LDAPSyncAction
+
+
+class BootstrapType(StrEnum):
+    """Represents the type of bootstrap operation."""
+
+    CREATION = "creation"
+    RESET = "reset"
+
+
+class BootstrapView(ui.View):
+    """View for the LDAP bootstrap command."""
+
+    def __init__(self, cog: "LDAP") -> None:
+        super().__init__(timeout=None)
+        self.cog = cog
+
+    @ui.button(
+        label="Create or Reset Login", style=discord.ButtonStyle.primary, custom_id="generate_creds"
+    )
+    async def generate_creds(self, interaction: discord.Interaction, _button: ui.Button) -> None:
+        """Generate credentials for the user."""
+        user = interaction.user
+
+        if LDAP_BASE_STAFF_ROLE not in [role.id for role in user.roles]:
+            await interaction.response.send_message(
+                "You are not eligible for LDAP enrollment.", ephemeral=True
+            )
+            return
+
+        bootstrap_type, password = await self.cog.bootstrap(user)
+
+        if bootstrap_type == BootstrapType.CREATION:
+            title = "Account Creation"
+        else:
+            title = "Password Reset"
+
+        content = CREDENTIALS_SECTION.format(title=title, username=user.name, password=password)
+
+        await interaction.response.send_message(content, ephemeral=True)
+
+
+class LDAP(commands.Cog):
+    """Commands for working with the LDAP Directory."""
+
+    def __init__(self, bot: KingArthur) -> None:
+        self.bot = bot
+        self.sync_users.start()
+
+    @tasks.loop(minutes=10)
+    async def sync_users(self) -> None:
+        """Sync users with the LDAP directory."""
+        logger.info("Syncing users with the LDAP directory.")
+
+        diff, missing_emp = await self.get_user_diff()
+
+        add_users = len([user for user in diff if user.action == LDAPSyncAction.ADD])
```

Comment:
This should also be possible (or `collections.Counter` if you want to get real fancy)
```py
user_actions = [user.action for user in diff]
add_users = user_actions.count(LDAPSyncAction.ADD)
```